File: OnlyCueTests/GoldenDoubleTests.swift
import XCTest

/// #833 — what `GoldenDouble`'s transport can and cannot carry.
///
/// The golden vectors compare doubles **bitwise**, so the text in the file has
/// to determine the bit pattern exactly. A mnemonic cannot: Swift's `"nan"`
/// means `0x7FF8000000000000` and .NET parses the same four characters back to
/// `0xFFF8000000000000` (its `double.NaN` carries the sign bit). Both were
/// measured, not assumed. So NaN travels as its bits, and every spelling only
/// one platform accepts is refused by both.
final class GoldenDoubleTests: XCTestCase {

    // MARK: - NaN carries its bits

    func test_theCanonicalNaN_travelsAsItsBitPattern() throws {
        XCTAssertEqual(try encode(.nan), "nan:0x7FF8000000000000")
    }

    /// The acceptance case from #833: a NaN that came out of an algorithm rather
    /// than from the `Double.nan` constant. Widening any binary32 NaN yields
    /// `0x7FF8…`, which is precisely the pattern .NET cannot name.
    func test_aWidenedBinary32NaN_spellsTheSameAsTheCanonicalNaN() throws {
        let widened = Double(Float(bitPattern: 0x7FC0_0000))
        XCTAssertEqual(try encode(widened), "nan:0x7FF8000000000000")
    }

    func test_theSignBitSurvives() throws {
        XCTAssertEqual(try encode(-Double.nan), "nan:0xFFF8000000000000")
        XCTAssertNotEqual(try decode("nan:0x7FF8000000000000"),
                          try decode("nan:0xFFF8000000000000"))
    }

    func test_aNaNPayloadSurvives() throws {
        let payload = Double(bitPattern: 0x7FF8_0000_0000_0001)
        XCTAssertEqual(try encode(payload), "nan:0x7FF8000000000001")
        XCTAssertEqual(try decode("nan:0x7FF8000000000001").bitPattern, payload.bitPattern)
    }

    // MARK: - Spellings only one platform accepts

    /// The mnemonics are refused rather than reinterpreted. Accepting them would
    /// mean picking a sign for the reader, and the two platforms pick different
    /// ones — which is the whole defect.
    func test_bareNaNMnemonics_areRejected() {
        for text in ["nan", "-nan", "NaN"] {
            XCTAssertThrowsError(try decode(text), "\(text) must not decode")
        }
    }

    /// `Double("0x7FF8000000000000")` is **9.221120237041091e+18** on Swift — a
    /// plausible finite number, silently — while .NET's `TryParse` refuses it.
    /// Both measured. A bare-hex spelling would therefore have been the worst
    /// possible carrier for a bit pattern: wrong on one side without failing,
    /// which is why the marker prefix exists.
    func test_hexFloatSpellings_areRejected() {
        for text in ["0x7FF8000000000000", "0x1p3", "-0x1p3", "0x1F"] {
            XCTAssertThrowsError(try decode(text), "\(text) must not decode")
        }
    }

    func test_aMalformedBitPattern_isRejected() {
        for text in ["nan:0x", "nan:0xZZZ8000000000000", "nan:7FF8000000000000",
                     "nan:0x7FF80000000000000"] {
            XCTAssertThrowsError(try decode(text), "\(text) must not decode")
        }
    }

    // MARK: - Everything else is unchanged

    /// The infinities have exactly one bit pattern each, so their mnemonics are
    /// lossless and stay. This pins that the NaN change did not disturb them.
    func test_finiteValuesAndInfinities_keepTheirSpelling() throws {
        for (value, text) in [(1.0, "1.0"), (-0.0, "-0.0"), (1e-05, "1e-05"),
                              (Double.infinity, "inf"), (-Double.infinity, "-inf")] {
            XCTAssertEqual(try encode(value), text)
        }
        for (text, value) in [("1.0", 1.0), ("-0.0", -0.0), ("1e-05", 1e-05),
                              ("inf", Double.infinity), ("-inf", -Double.infinity)] {
            XCTAssertEqual(try decode(text).bitPattern, value.bitPattern, text)
        }
    }

    // MARK: - Harness

    /// Encoded inside an array because the vectors never carry a bare fragment,
    /// and the quotes are stripped so the assertions read as the file's text.
    private func encode(_ value: Double) throws -> String {
        let data = try JSONEncoder().encode([GoldenDouble(value)])
        let text = try XCTUnwrap(String(data: data, encoding: .utf8))
        return String(text.dropFirst(2).dropLast(2))
    }

    private func decode(_ text: String) throws -> Double {
        let data = try XCTUnwrap("[\"\(text)\"]".data(using: .utf8))
        return try JSONDecoder().decode([GoldenDouble].self, from: data)[0].value
    }
}

File: windows/OnlyCue.Core.Tests/GoldenDoubleTests.cs
using Xunit;

namespace OnlyCue.Core.Tests;

/// <summary>
/// #833 — the reading half of what <c>OnlyCueTests/GoldenDoubleTests.swift</c>
/// pins on the writing side.
/// </summary>
/// <remarks>
/// The trap this closes, measured on both runtimes rather than assumed: Swift's
/// <c>Double.nan</c> is <c>0x7FF8000000000000</c> and so is every NaN produced by
/// widening a binary32 NaN, while .NET's <c>double.NaN</c> is
/// <c>0xFFF8000000000000</c> — and <c>double.TryParse</c> maps <i>both</i>
/// <c>"nan"</c> and <c>"-nan"</c> onto the sign-bit-set pattern, so no text
/// .NET understands can name the one Swift actually writes. Carrying the bits
/// is the only spelling that works in both directions.
/// </remarks>
public class GoldenDoubleTests
{
    /// <summary>The acceptance case from #833.</summary>
    [Fact]
    public void Parse_CarriesTheNaNSwiftActuallyWrites()
    {
        var widened = (double)BitConverter.Int32BitsToSingle(0x7FC00000);

        Assert.True(GoldenDouble.BitwiseEquals(GoldenDouble.Parse("nan:0x7FF8000000000000"), widened));
    }

    /// <summary>
    /// The other half of the acceptance: the two NaNs must stay distinguishable.
    /// Before the fix this comparison could not even be expressed — every NaN
    /// spelling collapsed onto <c>double.NaN</c>.
    /// </summary>
    [Fact]
    public void Parse_KeepsTheSignedNaNDistinctFromTheUnsignedOne()
    {
        var unsigned = GoldenDouble.Parse("nan:0x7FF8000000000000");
        var signed = GoldenDouble.Parse("nan:0xFFF8000000000000");

        Assert.False(GoldenDouble.BitwiseEquals(unsigned, signed));
        Assert.True(GoldenDouble.BitwiseEquals(signed, double.NaN));
    }

    [Fact]
    public void Parse_KeepsANaNPayload()
    {
        var parsed = GoldenDouble.Parse("nan:0x7FF8000000000001");

        Assert.Equal(0x7FF8000000000001, BitConverter.DoubleToInt64Bits(parsed));
    }

    /// <summary>
    /// Refused rather than reinterpreted. Accepting a mnemonic would mean
    /// choosing a sign on the writer's behalf, and the two platforms choose
    /// differently — which is the defect itself.
    /// </summary>
    [Theory]
    [InlineData("nan")]
    [InlineData("-nan")]
    [InlineData("NaN")]
    [InlineData("Infinity")]
    [InlineData("-Infinity")]
    public void Parse_RejectsSpellingsOnlyOnePlatformUnderstands(string text)
    {
        Assert.Throws<InvalidDataException>(() => GoldenDouble.Parse(text));
    }

    /// <summary>
    /// A regression guard, not a fix: .NET already refuses these. Swift does
    /// not — <c>Double("0x1p3")</c> is 8.0 there — so the Swift reader grew the
    /// matching rejection and this keeps the two from drifting back apart.
    /// </summary>
    [Theory]
    [InlineData("0x7FF8000000000000")]
    [InlineData("0x1p3")]
    [InlineData("0x1F")]
    public void Parse_RejectsHexFloatSpellings(string text)
    {
        Assert.Throws<InvalidDataException>(() => GoldenDouble.Parse(text));
    }

    /// <remarks>
    /// <c>"nan:0x7FF8"</c> earns its place: a mutation run showed every other
    /// case here still failing without the 16-digit length check, because they
    /// are all unparseable hex. A <em>short but valid</em> pattern is the one
    /// that goes quietly wrong — it reads back as <c>0x7FF8</c>, a denormal
    /// around 1.6e-319, which is a finite number where a NaN was meant.
    /// </remarks>
    [Theory]
    [InlineData("nan:0x")]
    [InlineData("nan:0x7FF8")]
    [InlineData("nan:0x0")]
    [InlineData("nan:0xZZZ8000000000000")]
    [InlineData("nan:7FF8000000000000")]
    [InlineData("nan:0x7FF80000000000000")]
    public void Parse_RejectsAMalformedBitPattern(string text)
    {
        Assert.Throws<InvalidDataException>(() => GoldenDouble.Parse(text));
    }

    /// <summary>The infinities have one bit pattern each, so their mnemonics are
    /// lossless and stay as they were. This pins that the NaN change left
    /// them — and the ordinary decimals — alone.</summary>
    [Theory]
    [InlineData("inf", 0x7FF0000000000000L)]
    [InlineData("-inf", unchecked((long)0xFFF0000000000000UL))]
    [InlineData("1.0", 0x3FF0000000000000L)]
    [InlineData("-0.0", unchecked((long)0x8000000000000000UL))]
    [InlineData("1e-05", 0x3EE4F8B588E368F1L)]
    public void Parse_LeavesEveryOtherSpellingUnchanged(string text, long bits)
    {
        Assert.Equal(bits, BitConverter.DoubleToInt64Bits(GoldenDouble.Parse(text)));
    }
}
